File: lib/controller/zone.dart
import 'dart:convert';

import 'package:cartracker_backend/database/zone_dao.dart';
import 'package:mongo_dart/mongo_dart.dart';
import 'package:shelf/shelf.dart';

class ZoneController {
  Future<Response> createZone(Request request) async {
    var body = jsonDecode(await request.readAsString());

    String id = Uuid().v4();
    String name = body["name"];
    double radius = body["radius"];
    List<dynamic> pos = body["pos"];

    List pos2 = []; // TODO List<Pos>
    for (var p in pos) {
      pos2.add(p);
    }

    if ((pos2.length == 1 && radius > 0) ||
        (pos2.length >= 3 && radius == 0)) {
      await ZoneDao.create(Zone(id: id, name: name, radius: radius, pos: pos2));

      return Response(201, body: jsonEncode(id));
    }
    return Response(400);
  }

  Future<Response> getZones(Request request) async {
    var res = await ZoneDao.readAll();
    return Response(200, body: jsonEncode(res));
  }

  Future<Response> getZoneByID(Request request, String id) async {
    return Response(200, body: jsonEncode(null));
  }

  Future<Response> updateZone(Request request, String id) async {
    var body = jsonDecode(await request.readAsString());

    String name = body["name"];
    double radius = body["radius"];
    List<dynamic> pos = body["pos"];

    List pos2 = []; // TODO List<Pos>
    for (var p in pos) {
      pos2.add(p);
    }

    if ((pos2.length == 1 && radius > 0) ||
        (pos2.length >= 3 && radius == 0)) {
      await ZoneDao.update(id, name, radius, pos2);
      return Response(200, body: jsonEncode(body));
    }
    return Response(400);
  }

  Future<Response> deleteZone(Request request, String id) async {
    await ZoneDao.delete(id);
    return Response(200, body: jsonEncode(id));
  }
}

File: src/app/models/zone.ts

export interface Zone {
  _id?: string,
  id?: string,
  name: string,
  pos: Pos[],
  radius: number
}

export interface Pos {
  lat: number,
  lng: number
}

export class ZoneClass implements Zone {
  public _id?: string;
  public id?: string;
  public name: string;
  public pos: Pos[];
  public radius: number;

  constructor(name: string, pos: Pos[], radius: number, _id?: string, id?: string) {
    this._id = _id;
    this.id = id;
    this.name = name;
    this.pos = pos;
    this.radius = radius;
  }
}

export class PosClass implements Pos {
  public lat: number;
  public lng: number;

  constructor(lat: number, lng: number) {
    this.lat = lat;
    this.lng = lng;
  }
}
